### ISSUE TITLE
Poor Feature/Example serialization performance

### ISSUE THREAD

**Issue by novog** [tensorflow/tensorflow#30372]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10 Enterprise
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): unknown 2.0.0-beta1
- Python version: 3.7.2
- CUDA/cuDNN version: 10.0/7.5
- GPU model and memory: GeForce GTX Titan X, 12GB

**Describe the current behavior**
The rate of serializing examples is unreasonably slow. In particular, performance is slow for `Feature`s that are a sequence of values. I have a dataset with numerous fields, including two that are represented by short, fixed length, 1-dimensional `Tensor`s. When excluding those two features, serialization happens in a slow but manageable amount of time. Adding either of those two features causes it to take many times as long.

Notably, when mapping a `Dataset` to a function that performs serialization, increasing the number of threads does not significantly impact performance and the CPU remains mostly idle. I do see marked performance improvements and higher CPU usage when parallelizing other mapped functions, so it could be that there is some kind of global bottleneck for this operation.

**Describe the expected behavior**
Records should be serialized in a reasonable amount of time.

**Code to reproduce the issue**
The following code serializes 10000 records in about 6s on a particular machine. Note that if I replace the mapped function with one that simply returns a constant, it takes less than 1s to complete, so the serialization is the problem.

    def make_example_1(*data_list):
        feature_dict = {
            'a':tf.train.Feature(int64_list=tf.train.Int64List(value=[data_list[0]])),
            'b':tf.train.Feature(int64_list=tf.train.Int64List(value=[data_list[1]])),
            'c':tf.train.Feature(int64_list=tf.train.Int64List(value=[data_list[2]])),
            'd':tf.train.Feature(int64_list=tf.train.Int64List(value=[data_list[3]])),
            'e':tf.train.Feature(int64_list=tf.train.Int64List(value=[data_list[4]])),
            'f':tf.train.Feature(int64_list=tf.train.Int64List(value=[data_list[5]])),
            'g':tf.train.Feature(int64_list=tf.train.Int64List(value=[data_list[6]])),
            'h':tf.train.Feature(int64_list=tf.train.Int64List(value=[data_list[7]])),
            'i':tf.train.Feature(int64_list=tf.train.Int64List(value=[data_list[8]])),
            'j':tf.train.Feature(int64_list=tf.train.Int64List(value=[data_list[9]]))}
        example = tf.train.Example(features=tf.train.Features(feature=feature_dict))
        return example.SerializeToString()
    
    def make_example_1_wrap(*input):
        data = input[0]
        return tf.py_function(
            make_example_1,
            (
                data['a'],data['b'],data['c'],data['d'],data['e'],
                data['f'],data['g'],data['h'],data['i'],data['j']),
            Tout=tf.string)
        
    def feature_test_1(num_threads=None):
        source = {
            'a':tf.constant(0),
            'b':tf.constant(1),
            'c':tf.constant(2),
            'd':tf.constant(3),
            'e':tf.constant(4),
            'f':tf.constant(5),
            'g':tf.constant(6),
            'h':tf.constant(7),
            'i':tf.constant(8),
            'j':tf.constant(9)}
        ds = tf.data.Dataset.from_tensors(source).repeat(10000)
        ds = ds.map(make_example_1_wrap, num_threads)
        it = iter(ds)
        for x in it:
            pass

The following example uses about the same input data size as the previous one, but uses 2 features of length 5 instead of 10 features of length 1. The execution time increases to 17s, highlighting the problem with sequences as `Int64List`s.

    def make_example_2(*data_list):
        feature_dict = {
            'a':tf.train.Feature(int64_list=tf.train.Int64List(value=data_list[0])),
            'b':tf.train.Feature(int64_list=tf.train.Int64List(value=data_list[1]))}
        example = tf.train.Example(features=tf.train.Features(feature=feature_dict))
        return example.SerializeToString()
    
    def make_example_2_wrap(*input):
        data = input[0]
        return tf.py_function(
            make_example_2,
            (data['a'], data['b']),
            Tout=tf.string)
    
    def feature_test_2(num_threads=None):
        x = {
            'a':tf.constant([0,1,2,3,4]),
            'b':tf.constant([5,6,7,8,9])}
        ds = tf.data.Dataset.from_tensors(x).repeat(10000)
        ds = ds.map(make_example_2_wrap, num_threads)
        it = iter(ds)
        for x in it:
            pass

**Other info / logs**
A possibly related performance issue is mentioned by many users in #16933, although that issue was closed over a year ago due to inactivity.

**Comment by areeh** [comment]
Yes, same issue as mentioned in other issues like #16933. To write something performant we've used `tf.serialize_tensor` with `tf.data.experimental.TFRecordWriter()` as mentioned briefly eg in the recent [load images](https://www.tensorflow.org/tutorials/load_data/images) tutorial. This is quite cumbersome to use compared to Examples and Features.

**Comment by rmothukuru** [comment]
@novog ,
Can you please confirm if using the APIs and referring the Tutorial mentioned by @areeh has given you expected performance.

**Comment by novog** [comment]
OK, thanks. If using `Example` isn't viable for the moment, then I suppose that will do.

In my case, I have not a single tensor, but a number of tensors corresponding to named features of differing types, so I can't just use `tf.io.serialize_tensor`. I could serialize them individually, put them in a dictionary, serialize the dictionary (using something like python's `json`), then pass those strings to `TFRecordWriter`. Let me know if you have a cleaner solution. Otherwise, I'll try that and post a follow-up on the relative performance.

**Comment by novog** [comment]
Yes, it is much faster using `tf.io.serialize_tensor` manually rather than using `Feature`/`Example`. The following code completes in about 3.5s instead of 17s:

    def make_example_2(*data_list):
        feature_dict = {
            'a':tf.io.serialize_tensor(data_list[0]).numpy(),
            'b':tf.io.serialize_tensor(data_list[1]).numpy()}
        serialized = pickle.dumps(feature_dict)
        return serialized

Given that mapping to a function that simply returns a constant string takes 2.5s, the time added by serialization decreased from about 14.5s to 1s.

As to the status of this issue, while there is a sort of workaround, I would consider the defect itself to remain: `Feature`/`Example` perform unreasonably poorly.

**Comment by areeh** [comment]
@novog Yes and I don't know of a way to have behaviour more or less equivalent to `Feature`/`Example` with `serialize_tensor`

**Comment by alextp** [comment]
I don't think we can meaningfully speed up protocol buffer serialization in TF. I suggest you follow this bug up with one against the maintaners of protocol buffers themselves.

Either that or use, as suggested above, another serialization format.

**Comment by tracek** [comment]
I tried [both approaches](https://stackoverflow.com/questions/62174662/slow-serialization-of-ndarray-to-tfrecord) and frankly `serialize_tensor` performed even worse. Ideas?

**Comment by xykong** [comment]
> Yes, it is much faster using `tf.io.serialize_tensor` manually rather than using `Feature`/`Example`. The following code completes in about 3.5s instead of 17s:
> 
> ```
> def make_example_2(*data_list):
>     feature_dict = {
>         'a':tf.io.serialize_tensor(data_list[0]).numpy(),
>         'b':tf.io.serialize_tensor(data_list[1]).numpy()}
>     serialized = pickle.dumps(feature_dict)
>     return serialized
> ```
> 
> Given that mapping to a function that simply returns a constant string takes 2.5s, the time added by serialization decreased from about 14.5s to 1s.
> 
> As to the status of this issue, while there is a sort of workaround, I would consider the defect itself to remain: `Feature`/`Example` perform unreasonably poorly.

Hello @novog , I tried this idea, It's much faster than tf.train.Feature. But I can't make the read pipeline work.
When I call tf.io.parse_tensor, it's failed with the error message 
2021-07-03 23:35:01.655594: W tensorflow/core/framework/op_kernel.cc:1755] Invalid argument: ValueError: Tensor conversion requested dtype string for Tensor with dtype int32: <tf.Tensor: shape=(9, 9, 395), dtype=int32, numpy=

could you please show me the code how to read from the make_example_2 ?

**Comment by novog** [comment]
@xykong If I understand your question correctly, you are trying to use `parse_tensor` to transform the serialized string that we created back to a tensor. Instead, you would need to call `pickle.loads` on the string after reading it in from disk. Doing so will yield the `feature_dict` object that we had serialized.

In any case, I believe that the general movement has been toward using Keras functionality in place of things like Feature/FeatureColumn/Example, so you might want to check out how Keras preprocessing layers and data generators work and consider them as an option.
